### ISSUE TITLE
Wish: Standard Pane Name "Default" - possibly phase out the Pane "Admin"

### ISSUE THREAD

**Issue by iJungleboy** [oqtane/oqtane.framework#2382]
ATM there is only one standard Pane name called "Admin".

This is weird, because any user adding controls to "Admin" will feel like it's not right. 

At the same time, there is no official name for any other pane, and I believe "Default" would be the best option. 

Since there is no official best-name, Themes can pick anything they want, but this makes it really hard to switch themes, as the names won't match.

My recommendation

1. Introduce a standard `PaneNames.Default = "Default"`
2. Encourage the use of this as opposed to `Admin`
3. ...possibly phase out `Admin` or change it so it's only visible to admins anyhow
4. ...possibly auto-fallback to put stuff which was in `Admin` to `Default`

**Comment by sbwalker** [comment]
There is one required pane in an Oqtane theme and it is called "Admin". The name "Admin" was chosen because it is only administrators who can actually see the structure of a page... and also because Oqtane took a different approach from DNN in this area and rather than requiring 2 themes - an "admin" theme as well as a "page" theme - Oqtane consolidated these into a single concept - the Admin pane replaced the need for an Admin theme. It is certainly debatable on whether this is the best name for the default pane, however I do not view this as a high priority at this time.

**Comment by iJungleboy** [comment]
@sbwalker I understand that thought and for admin purpose it makes sense.

The problem is that the initial content in the normal theme is also in the `Admin` pane. 
This in turn will cause most developers to assume that the `Admin` pane is the default pane and should be used as such. Simply because they see a convention and assume this is how it should be done. So I assume almost nobody will question this, and that in turn will result in all newly created themes assuming that `Admin` = `Content` = `Default`. 

Anybody else's thoughts on this?

**Comment by sbwalker** [comment]
The Admin pane is the default pane. It is the only required pane in a theme.

**Comment by iJungleboy** [comment]
I totally agree that `Default` would be preferrable to `Content`. 

And the localization of the default pane names should happen at global level ideally, but the theme creator should be able to provide his own localization if he/she opts to using many themes.

**Comment by sbwalker** [comment]
In theory the existing framework abstraction should have made this very simple - just by changing the value of PaneNames.Admin all affected areas would be covered (the framework as well as the default themes distributed with the framework already use this constant for precisely this reason):

```
namespace Oqtane.Shared {
    public class PaneNames {
        public const string Admin = "Default";
    }
}
```
However, this is where theory does not match reality. Because this is a constant, the value is embedded into any assembly which references it. So changes to the constant value are not propagated to other assemblies unless they are explicitly recompiled against the newer version of the assembly where the constant value is defined. If this constant had instead been defined as a static readonly then it would have been resolved at run-time and the change would have been simple.

So this gets a lot more complicated... as the framework would now need to support 2 "default" panes, with fallback support, potential issues if someone declared both panes in their theme, backward compatibility logic, etc... I do not see the benefits outweighing challenges of going down this path. 

Alternatively, we do need a localization solution for panes. For example, if you consider the fact that the list of panes needs to be declared in your theme component:

```
    public override string Panes => PaneNames.Admin + ",Top Full Width,Top 100%,Left 50%,Right 50%,Left 33%,Center 33%,Right 33%,Left Outer 25%,Left Inner 25%,Right Inner 25%,Right Outer 25%,Left 25%,Center 50%,Right 25%,Left Sidebar 66%,Right Sidebar 33%,Left Sidebar 33%,Right Sidebar 66%,Bottom 100%,Bottom Full Width,Footer";
```

There needs to be a way to include a *.resx file with your theme which localizes this value. If this capability existed then you could include resources for the default language (English) which localizes "Admin" to "Default". With this approach the framework continues to use PaneNames.Admin but the end user sees "Default Pane" in the UI.

**Comment by iJungleboy** [comment]
The static vs. const issue is a real problem, totally agree. I'm glad we fixed it elsewhere (I think it was the version or something). 

But just renaming it wouldn't actually work, because the DB contains all the string-references with this term. I've just opened #2386 because I see this as a related, big issue. 

But if we do fix #2386, then I believe it would fix this issue here automatically, and we could move to `Default` or not - either way it would work.

**Comment by sbwalker** [comment]
You are correct that the PageModule table contains a Pane column which identifies the pane where a module should be injected on a page - which further complicates making a change from Admin to Default as some themes may use the old constant and others the new constant - which makes things very messy. 

Which basically brings me back to my original comment that changing the name of the default pane has more disadvantages than advantages and therefore is not a priority. I believe the alternative solution of providing localization support is the better option here.

**Comment by iJungleboy** [comment]
@sbwalker I understand.

Basically if we get #2386 to work this would become obsolete anyhow, because in that case `Admin` is the default, but any theme that choses to only have a `Default`, `Content` or `Standard` would still work anyhow. So if we could do that, then this issue would become irrelevant. 

So based on that if you agree, we could close this issue.

**Comment by sbwalker** [comment]
There is actually a bit more to this story...

Oqtane has 2 distinct page rendering modes:

1. Standard page routes - these are standard pages which can have many different modules rendered into a variety of different panes defined in a theme. By default it is the Index component of a module which is rendered into panes in this mode. The "admin" pane is not important to this rendering mode. 

2. Page routes with an action and module id parameter - these are "admin" pages which are only rendered for users who have edit or administrative rights. In this mode only a single module component defined by the action parameter is rendered into the "admin" pane of the page. This is where the "admin" pane plays a critical role as depending on how you have defined your admin container behavior (ie. it does not need to be a modal dialog) this pane needs to be sized to provide a good administrative user experience. 

So getting back to your issue... the Control Panel allows you to select a specific pane from a list of panes defined for the theme. The list is ordered based on the order defined in the Panes property within the theme component. The first time an administrator uses the Control Panel, the first pane in the list will be set as the default. This happens to be the Admin pane in both of the default themes in the framework currently because that is the way the Panes value was defined. But this could easily be changed to specify the Admin pane at the end of the Panes list. Note that once an administrator chooses a pane the framework will remember their choice as a default for the next time they add a module to a page. So this simple change to the Panes property in the Oqtane default theme as well as your own themes could solve one of the issues you identified above. 

The default site template assigns the content to the Admin pane. It could be modified to place content in a different pane. The issue is that a theme such as the default Blazor theme only defines a single pane - the Admin pane. This is because it does not actually need any other panes as the theme is only intended for hosting Blazor apps - not content. And since the only "required" pane in a theme is the Admin pane it does not make sense to change the default site template. 

In light of mode #2 above I am rethinking what I said in the other issue you logged. It is important for developers to follow the convention and include an Admin pane. Including a fallback to the first pane in the list will end up causing issues so I am leaning more towards displaying an error dialog now.

**Comment by sbwalker** [comment]
@iJungleboy In an effort to try to reconcile your request with the requirements of the framework, I believe I have a workable solution:

1. The Admin pane needs to be retained as it serves a critical role as outline in #2 above. It will remain as the one required pane in every theme
2. Added validation so that the user is alerted if they try to use a theme which either has no panes defined or does not have an Admin pane defined
3. Modified Control Panel so that it selects the first pane in the list by default for a theme (it was selecting the Admin pane previously)
4. If a module is configured to inject into a pane which does not exist in a theme, it will fallback to use the Admin pane (same behavior as previously)
5. Introduced a PaneNames.Default constant. It is available for use in themes but is not a requirement as nothing actually depends on it. 
6. Updated the Oqtane theme to include a Default pane (the Blazor theme will continue to only have only an Admin pane). The Default pane is listed first. The Admin pane is now listed last.
7. Updated the site template logic to place modules in the Default pane for all new installations.

There should be no compatibility issues as a result of these changes as existing installations will continue to inject modules into themes based on the specification in the database. New installations which use themes created prior to these changes will use the fallback ie. modules intended to be injected in the Default pane will be injected in the Admin pane as they were in the past which provides the same experience. Items 5,6,7, are not critical but should help alleviate the confusion for users related to the Admin pane name in the future.

**Comment by sbwalker** [comment]
4. above could be modified so that if a module is configured to inject into a pane which does not exist in a theme it could check if the theme contains a Default pane and if so, use it... and if not, use the Admin pane. This would make the Default pane play more of a prominent role in the framework (although still not a hard requirement).

**Comment by sbwalker** [comment]
After all of the discussion it seems that there are 2 distinct scenarios:

1. Page routes with an action and module id parameter - this requires the "Admin" pane so it is a framework requirement
2. Standard page routes - this could benefit from a "Default" pane from a user experience and theme migration perspective. Perhaps it should be a soft "requirement" to include one. Legacy themes that do not have a  Default pane can fallback to using the Admin pane.

Thoughts?

**Comment by iJungleboy** [comment]
@sbwalker this sounds pretty good, yet I would ask for some tweaking

1. I would default to rescuing modules into the `Default` pane first `Admin` afterwards if this is feasible. If it's not, then `Admin` first,  `Default` last as you mentioned later.
1. I would not warn if admin-pane is missing, except in the cases where it's required. This would be important to solutions with strong editing focus. Fo example, we don't want to have an `Admin` pane in most normal themes/sites as they confuse the editor and would possibly result in different styling which again confuses the editor. For example, our current workaround is to actually have the pane `Admin` but not mention it in the list of panes, because that way the editor can't accidentally put things in it. After this change we would actually hide the `Admin` pane on all pages using an `@if (...)` unless we're in a scenario requiring it.
2. Possibly providing a theme property like `IsAdminPage` or something to indicate that we're in such a routing-scenario would be helpful as well, for themes showing/hiding the admin pane like we would like to do. 

sound good?

**Comment by sbwalker** [comment]
Responses below:

1. Yes it could fallback to Default pane first if it exists... and if not then Admin pane. Please note that I am only referring to standard page routes here - not the route scenario described below. 

2. The Admin pane must exist in every theme - it is used in the route scenario I described earlier in this thread where the Url contains an action and module id parameter. To provide a concrete example, when an Html/Text is on a page and an authorized user clicks the Edit button for the module - the page is re-rendered with only the edit.razor component (the rich text editor) displayed in the Admin pane. This is the way that every administrative module in the Admin Panel functions as well (ie. edit page, edit role, edit user, etc...). So the Admin pane is absolutely required in every theme - and if a theme does not contain an Admim pane the user needs to be alerted. 

3. Don't understand. I need to understand if you are trying to avoid the Admin simply because you don't like the name, or because you don't like the functionality. To be more clear do you believe that a theme should only have 1 required pane and it should be named Default and it should be used in all route scenarios (basically exactly the way that the Admin pane was operating prior to any recent changes - minus the name of course)?

I am concerned that we are making the overall solution more complicated - and complicated is the enemy of good. This needs to be as simple as possible while still provoking a good user experience.

**Comment by sbwalker** [comment]
Ok, final solution...

1. There is 1 "required" pane in a theme
2. The required pane can either be "Default" (new recommended option) or "Admin" (legacy)
3. Framework will validate that a theme declares its list of Panes and that the list contains either a Default or Admin pane
4. Page route will use the Default pane as the fallback if a theme contain it, otherwise Admin pane
5. Action route will use the Default pane as the fallback if a theme contain it, otherwise Admin pane
6. Control Panel loads the list of pane options from the Panes list defined for a theme - so no conditional logic is needed to show/suppress the Admin pane option
7. Control Panel will use the Default pane as the default option in the list if a theme contain it, otherwise Admin pane (rather tha first pane in pane list)
8. Oqtane Theme will include a Default pane and not an Admin pane
9. Blazor Theme will include only an Admin pane (to prove backward compatibility)

These changes are included in #2400

**Comment by iJungleboy** [comment]
This is really, really awesome. 
Thanks.
You 🤘🏾!
